File: BioLLM_RawDataAnalysis/frontend/src/workflow-editor/validation.ts
import type { NodeRegistry } from './registry'
import type {
  WorkflowDocument,
  WorkflowEdge,
  WorkflowValidationIssue,
  WorkflowValidationResult,
} from './types'

function issue(
  severity: WorkflowValidationIssue['severity'],
  code: string,
  message: string,
  nodeIds: string[] = [],
  edgeId?: string,
  confirmationKey?: string,
): WorkflowValidationIssue {
  return {
    severity,
    code,
    message,
    node_ids: nodeIds,
    edge_id: edgeId,
    confirmation_key: confirmationKey,
  }
}

function topologicalOrder(workflow: WorkflowDocument): { order: string[]; cyclic: boolean } {
  const ids = workflow.nodes.map(node => node.id)
  const known = new Set(ids)
  const incoming = new Map(ids.map(id => [id, 0]))
  const outgoing = new Map(ids.map(id => [id, [] as string[]]))
  workflow.edges.forEach(edge => {
    if (!known.has(edge.source_node) || !known.has(edge.target_node)) return
    incoming.set(edge.target_node, (incoming.get(edge.target_node) ?? 0) + 1)
    outgoing.get(edge.source_node)?.push(edge.target_node)
  })
  const ready = ids.filter(id => incoming.get(id) === 0)
  const order: string[] = []
  while (ready.length) {
    const current = ready.shift()!
    order.push(current)
    outgoing.get(current)?.forEach(target => {
      const count = (incoming.get(target) ?? 1) - 1
      incoming.set(target, count)
      if (count === 0) ready.push(target)
    })
  }
  return { order, cyclic: order.length !== ids.length }
}

export function connectionIssue(
  edge: WorkflowEdge,
  workflow: WorkflowDocument,
  registry: NodeRegistry,
): WorkflowValidationIssue | null {
  const source = workflow.nodes.find(node => node.id === edge.source_node)
  const target = workflow.nodes.find(node => node.id === edge.target_node)
  if (!source || !target) return issue('hard_error', 'unknown_edge_node', '连线引用了不存在的节点。', [edge.source_node, edge.target_node], edge.id)
  const sourceDefinition = registry.get(source.type)
  const targetDefinition = registry.get(target.type)
  const output = sourceDefinition?.outputs.find(port => port.id === edge.source_port)
  const input = targetDefinition?.inputs.find(port => port.id === edge.target_port)
  if (!output || !input) return issue('hard_error', 'unknown_port', '连线使用了注册表中不存在的端口。', [source.id, target.id], edge.id)
  if (!input.accepted_data_types.includes(output.data_type)) {
    return issue('hard_error', 'incompatible_port_types', `${output.data_type} 不能连接到 ${targetDefinition?.label}.${input.id}。`, [source.id, target.id], edge.id)
  }
  if (!input.accepted_scopes.includes(output.scope)) {
    return issue('hard_error', 'incompatible_data_scope', `${output.scope} 范围数据不能连接到 ${targetDefinition?.label}.${input.id}。`, [source.id, target.id], edge.id)
  }
  return null
}

export function validateWorkflow(workflow: WorkflowDocument, registry: NodeRegistry): WorkflowValidationResult {
  const issues: WorkflowValidationIssue[] = []
  const nodeIds = new Set<string>()
  workflow.nodes.forEach(node => {
    if (nodeIds.has(node.id)) issues.push(issue('hard_error', 'duplicate_node_id', '节点 ID 必须唯一。', [node.id]))
    nodeIds.add(node.id)
    if (!registry.has(node.type)) issues.push(issue('hard_error', 'unknown_node_type', `未注册的节点类型：${node.type}`, [node.id]))
  })

  const edgeIds = new Set<string>()
  const connected = new Map<string, number>()
  workflow.edges.forEach(edge => {
    if (edgeIds.has(edge.id)) issues.push(issue('hard_error', 'duplicate_edge_id', '连线 ID 必须唯一。', [], edge.id))
    edgeIds.add(edge.id)
    const compatibility = connectionIssue(edge, workflow, registry)
    if (compatibility) issues.push(compatibility)
    const key = `${edge.target_node}:${edge.target_port}`
    connected.set(key, (connected.get(key) ?? 0) + 1)
  })

  workflow.nodes.forEach(node => {
    const definition = registry.get(node.type)
    definition?.inputs.forEach(input => {
      const count = connected.get(`${node.id}:${input.id}`) ?? 0
      if (input.required && count === 0) issues.push(issue('hard_error', 'missing_required_input', `${definition.label}.${input.id} 需要上游连线。`, [node.id]))
      if (!input.multiple && count > 1) issues.push(issue('hard_error', 'multiple_connections_to_single_input', `${definition.label}.${input.id} 只允许一条输入连线。`, [node.id]))
    })
  })

  const topology = topologicalOrder(workflow)
  if (topology.cyclic) issues.push(issue('hard_error', 'cycle_detected', '工作流不能包含环路。'))

  // These are the backend's two explicitly allowed risk cases. They live only
  // in this validation adapter; components consume issues and never repeat rules.
  workflow.edges.forEach(edge => {
    const source = workflow.nodes.find(node => node.id === edge.source_node)
    const target = workflow.nodes.find(node => node.id === edge.target_node)
    const output = source && registry.get(source.type)?.outputs.find(port => port.id === edge.source_port)
    if (!source || !target || !output || edge.target_port !== 'reads') return
    let code: string | null = null
    let message = ''
    if (target.type === 'host_depletion' && output.data_type === 'paired_raw_reads') {
      code = 'host_depletion_without_fastp'
      message = '去宿主节点正在直接接收未经 fastp 清洗的原始 reads。'
    } else if (
      ['taxonomy', 'functional_annotation'].includes(target.type)
      && ['paired_raw_reads', 'paired_clean_reads'].includes(output.data_type)
    ) {
      code = 'annotation_without_host_depletion'
      message = '注释分析正在使用未去宿主的 reads，可能影响结果。'
    }
    if (code) {
      const key = `risk:${target.id}:${code}`
      issues.push({
        ...issue('warning', code, message, [target.id], edge.id, key),
        confirmed: workflow.accepted_risks.includes(key),
      })
    }
  })

  const hardErrors = issues.filter(item => item.severity === 'hard_error')
  const unconfirmed = issues.filter(item => item.severity === 'warning' && !item.confirmed)
  return {
    structurally_valid: hardErrors.length === 0,
    can_execute: hardErrors.length === 0 && unconfirmed.length === 0,
    topological_order: topology.order,
    issues,
  }
}

File: BioLLM_RawDataAnalysis/backend/app/workflows/validation.py
from __future__ import annotations

from collections import deque
from typing import Literal

from pydantic import BaseModel, Field

from .graph import WorkflowEdge, WorkflowGraph
from .registry import NodeRegistry


class GraphValidationIssue(BaseModel):
    severity: Literal["hard_error", "warning"]
    code: str
    message: str
    node_ids: list[str] = Field(default_factory=list)
    edge_id: str | None = None
    confirmation_key: str | None = None
    confirmed: bool = False


class GraphValidationResult(BaseModel):
    structurally_valid: bool
    can_execute: bool
    topological_order: list[str]
    issues: list[GraphValidationIssue]

    @property
    def hard_errors(self) -> list[GraphValidationIssue]:
        return [issue for issue in self.issues if issue.severity == "hard_error"]

    @property
    def warnings(self) -> list[GraphValidationIssue]:
        return [issue for issue in self.issues if issue.severity == "warning"]

    @property
    def unconfirmed_warnings(self) -> list[GraphValidationIssue]:
        return [issue for issue in self.warnings if not issue.confirmed]


def _hard_error(
    code: str,
    message: str,
    *,
    node_ids: list[str] | None = None,
    edge_id: str | None = None,
) -> GraphValidationIssue:
    return GraphValidationIssue(
        severity="hard_error",
        code=code,
        message=message,
        node_ids=node_ids or [],
        edge_id=edge_id,
    )


def _warning(
    code: str,
    message: str,
    *,
    node_id: str,
    accepted_risks: set[str],
    edge_id: str | None = None,
) -> GraphValidationIssue:
    key = f"risk:{node_id}:{code}"
    return GraphValidationIssue(
        severity="warning",
        code=code,
        message=message,
        node_ids=[node_id],
        edge_id=edge_id,
        confirmation_key=key,
        confirmed=key in accepted_risks,
    )


def _topological_order(
    node_ids: list[str], edges: list[WorkflowEdge]
) -> tuple[list[str], bool]:
    known = set(node_ids)
    incoming = {node_id: 0 for node_id in node_ids}
    outgoing = {node_id: [] for node_id in node_ids}
    for edge in edges:
        if edge.source_node not in known or edge.target_node not in known:
            continue
        outgoing[edge.source_node].append(edge.target_node)
        incoming[edge.target_node] += 1
    ready = deque(node_id for node_id in node_ids if incoming[node_id] == 0)
    ordered: list[str] = []
    while ready:
        current = ready.popleft()
        ordered.append(current)
        for target in outgoing[current]:
            incoming[target] -= 1
            if incoming[target] == 0:
                ready.append(target)
    return ordered, len(ordered) != len(node_ids)


def validate_workflow_graph(
    graph: WorkflowGraph, registry: NodeRegistry
) -> GraphValidationResult:
    issues: list[GraphValidationIssue] = []
    accepted_risks = set(graph.accepted_risks)
    nodes_by_id = {node.id: node for node in graph.nodes}

    if len(nodes_by_id) != len(graph.nodes):
        issues.append(
            _hard_error(
                "duplicate_node_id", "workflow node IDs must be unique"
            )
        )
    edge_ids = [edge.id for edge in graph.edges]
    if len(set(edge_ids)) != len(edge_ids):
        issues.append(
            _hard_error(
                "duplicate_edge_id", "workflow edge IDs must be unique"
            )
        )

    for node in graph.nodes:
        if node.type not in registry:
            issues.append(
                _hard_error(
                    "unknown_node_type",
                    f"unknown workflow node type: {node.type}",
                    node_ids=[node.id],
                )
            )

    connected_inputs: set[tuple[str, str]] = set()
    connection_counts: dict[tuple[str, str], int] = {}
    incoming_data_types: dict[tuple[str, str], str] = {}
    for edge in graph.edges:
        source = nodes_by_id.get(edge.source_node)
        target = nodes_by_id.get(edge.target_node)
        if source is None or target is None:
            issues.append(
                _hard_error(
                    "unknown_edge_node",
                    "edge references a node that is not present",
                    node_ids=[edge.source_node, edge.target_node],
                    edge_id=edge.id,
                )
            )
            continue
        source_definition = registry.get(source.type)
        target_definition = registry.get(target.type)
        if source_definition is None or target_definition is None:
            continue
        output_port = source_definition.output(edge.source_port)
        input_port = target_definition.input(edge.target_port)
        if output_port is None:
            issues.append(
                _hard_error(
                    "unknown_source_port",
                    f"node {source.id} has no output port {edge.source_port}",
                    node_ids=[source.id],
                    edge_id=edge.id,
                )
            )
            continue
        if input_port is None:
            issues.append(
                _hard_error(
                    "unknown_target_port",
                    f"node {target.id} has no input port {edge.target_port}",
                    node_ids=[target.id],
                    edge_id=edge.id,
                )
            )
            continue

        input_key = (target.id, input_port.id)
        connected_inputs.add(input_key)
        connection_counts[input_key] = connection_counts.get(input_key, 0) + 1
        incoming_data_types[input_key] = output_port.data_type
        if output_port.data_type not in input_port.accepted_data_types:
            issues.append(
                _hard_error(
                    "incompatible_port_types",
                    f"{output_port.data_type} cannot connect to {target.id}.{input_port.id}",
                    node_ids=[source.id, target.id],
                    edge_id=edge.id,
                )
            )
        if output_port.scope not in input_port.accepted_scopes:
            issues.append(
                _hard_error(
                    "incompatible_data_scope",
                    f"{output_port.scope} output cannot connect to {target.id}.{input_port.id}",
                    node_ids=[source.id, target.id],
                    edge_id=edge.id,
                )
            )
        if connection_counts[input_key] > 1 and not input_port.multiple:
            issues.append(
                _hard_error(
                    "multiple_connections_to_single_input",
                    f"{target.id}.{input_port.id} accepts only one connection",
                    node_ids=[target.id],
                    edge_id=edge.id,
                )
            )

    for node in graph.nodes:
        definition = registry.get(node.type)
        if definition is None:
            continue
        for input_port in definition.inputs:
            if input_port.required and (node.id, input_port.id) not in connected_inputs:
                issues.append(
                    _hard_error(
                        "missing_required_input",
                        f"{node.id}.{input_port.id} requires an upstream connection",
                        node_ids=[node.id],
                    )
                )

    topological_order, has_cycle = _topological_order(
        [node.id for node in graph.nodes], graph.edges
    )
    if has_cycle:
        issues.append(
            _hard_error("cycle_detected", "workflow graph must not contain cycles")
        )

    for node in graph.nodes:
        definition = registry.get(node.type)
        if definition is None:
            continue
        read_type = incoming_data_types.get((node.id, "reads"))
        incoming_edge = next(
            (
                edge
                for edge in graph.edges
                if edge.target_node == node.id and edge.target_port == "reads"
            ),
            None,
        )
        if node.type == "host_depletion" and read_type == "paired_raw_reads":
            issues.append(
                _warning(
                    "host_depletion_without_fastp",
                    "host depletion is receiving raw reads without fastp cleaning",
                    node_id=node.id,
                    accepted_risks=accepted_risks,
                    edge_id=incoming_edge.id if incoming_edge else None,
                )
            )
        if node.type in {"taxonomy", "functional_annotation"} and read_type in {
            "paired_raw_reads",
            "paired_clean_reads",
        }:
            issues.append(
                _warning(
                    "annotation_without_host_depletion",
                    "annotation is running without host-depleted reads",
                    node_id=node.id,
                    accepted_risks=accepted_risks,
                    edge_id=incoming_edge.id if incoming_edge else None,
                )
            )

    hard_errors = [issue for issue in issues if issue.severity == "hard_error"]
    unconfirmed = [
        issue
        for issue in issues
        if issue.severity == "warning" and not issue.confirmed
    ]
    structurally_valid = not hard_errors
    return GraphValidationResult(
        structurally_valid=structurally_valid,
        can_execute=structurally_valid and not unconfirmed,
        topological_order=topological_order,
        issues=issues,
    )
